WalletWasabi/Services/UpdateManager.cs

```diff
@@ -0,0 +1,238 @@
+		}
+		else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+		{
+			var cpu = RuntimeInformation.ProcessArchitecture;
+			if (cpu.ToString() == "Arm64")
+			{
+				var arm64url = urls.Where(url => url.Contains("arm64.dmg")).First();
+				return (arm64url, arm64url.Split("/").Last());
+			}
+			var url = urls.Where(url => url.Contains(".dmg") && !url.Contains("arm64")).First();
+			return (url, url.Split("/").Last());
+		}
+		else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+		{
+			throw new InvalidOperationException("For Linux, get the correct update manually.");
+		}
+		else
+		{
+			throw new InvalidOperationException("OS not recognized, download manually.");
+		}
+	}
+
+	public event EventHandler<UpdateStatus>? UpdateAvailableToGet;
+
+	public string DownloadsDir { get; }
+	public IHttpClient HttpClient { get; }
+
+	///<summary> Comes from config file. Decides Wasabi should download the new installer in the background or not.</summary>
+	public bool DownloadNewVersion { get; set; }
+
+	///<summary> Install new version on shutdown or not.</summary>
+	public bool DoUpdateOnClose { get; set; } = false;
+
+	public void InstallNewVersion()
+	{
+		try
+		{
+			var installerPath = Path.Combine(DownloadsDir, InstallerName);
+			ProcessStartInfo startInfo;
+			if (!File.Exists(installerPath))
+			{
+				throw new FileNotFoundException(installerPath);
+			}
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+			{
+				startInfo = ProcessStartInfoFactory.Make(installerPath, "", true);
+			}
+			else
+			{
+				startInfo = new()
+				{
+					FileName = installerPath,
+					UseShellExecute = true,
+					WindowStyle = ProcessWindowStyle.Normal
+				};
+			}
+
+			using Process? p = Process.Start(startInfo);
+
+			if (p is null)
+			{
+				throw new InvalidOperationException($"Can't start {nameof(p)} {startInfo.FileName}.");
+			}
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+			{
+				// For MacOS, you need to start the process twice, first start => permission denied
+				// TODO: find out why and fix.
+
+				p!.WaitForExit(5000);
+				p.Start();
+			}
+		}
+		catch (Exception ex)
+		{
+			Logger.LogError("Failed to install latest release. File might be corrupted.", ex);
+		}
+	}
+
+	public bool CheckIfInstallerDownloaded()
+	{
+		var folder = new DirectoryInfo(DownloadsDir);
+		if (folder.Exists)
+		{
+			FileSystemInfo[] files = folder.GetFileSystemInfos();
+
+			FileSystemInfo? file = files.Where(file => file.Name.Contains("Wasabi")).FirstOrDefault();
+
+			if (file is null)
+			{
+				Logger.LogDebug("Installer not found.");
+			}
+			else if (file.Name.Contains("tmp"))
+			{
+				Logger.LogWarning("Corrupted/unfinished installer found, deleting.");
+				File.Delete(file.FullName);
+			}
+			else
+			{
+				InstallerName = file.Name;
+				return true;
+			}
+
+			return false;
+		}
+		else
+		{
+			IoHelpers.EnsureDirectoryExists(DownloadsDir);
+			return false;
+		}
+	}
+
+	public void DeletePossibleLefotver()
```

Comment:
This method seems to be a duplicate for me. You are doing cleaning elsewhere too.

Reply:
When you update Wasabi, `GetInstallerAsync` won't run anymore, so corrupted or used files won't be deleted. This only runs when the `UpdateChecker` sees that the client is up to date, so this will delete the leftover installer that will not be used anymore.

Reply:
You can do this in UpdateStatusChangedAsync and keep the cleanings internal inside UpdateChecker.